compiler/src/iree/compiler/Codegen/Common/GPU/GPUPropagateDispatchSizeBounds.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-propagate-dispatch-size-bounds"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUPROPAGATEDISPATCHSIZEBOUNDSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+static void applyBounds(FunctionOpInterface funcOp,
+                        ArrayRef<int32_t> workgroupSizes,
+                        ArrayRef<int32_t> workgroupCounts) {
+  Builder b(funcOp->getContext());
+  funcOp->walk([&](Operation *op) {
+    TypeSwitch<Operation *>(op)
+        .Case([&](gpu::ThreadIdOp tidOp) {
+          tidOp.setUpperBoundAttr(b.getIndexAttr(
+              workgroupSizes[static_cast<uint32_t>(tidOp.getDimension())]));
+        })
+        .Case([&](IREE::HAL::InterfaceWorkgroupSizeOp wgSizeOp) {
+          wgSizeOp.setUpperBoundAttr(b.getIndexAttr(
+              workgroupSizes[wgSizeOp.getDimension().getZExtValue()]));
+        })
+        .Case([&](IREE::HAL::InterfaceWorkgroupIDOp wgIdOp) {
+          wgIdOp.setUpperBoundAttr(b.getIndexAttr(
+              workgroupCounts[wgIdOp.getDimension().getZExtValue()]));
+        })
+        .Case([&](IREE::HAL::InterfaceWorkgroupCountOp wgCountOp) {
+          wgCountOp.setUpperBoundAttr(b.getIndexAttr(
+              workgroupCounts[wgCountOp.getDimension().getZExtValue()]));
+        })
+        .Default([](Operation *) {});
+  });
+}
+
+struct GPUPropagateDispatchSizeBoundsPass final
+    : impl::GPUPropagateDispatchSizeBoundsPassBase<
+          GPUPropagateDispatchSizeBoundsPass> {
+  using GPUPropagateDispatchSizeBoundsPassBase::
+      GPUPropagateDispatchSizeBoundsPassBase;
```

Comment:
I think this should also work
```suggestion
  using Base::Base;
```